### ISSUE TITLE
Introduce EDD_Payment class

### ISSUE THREAD

**Issue by pippinsplugins** [easydigitaldownloads/Easy-Digital-Downloads#2667]
Creating payments and retrieving payment data can be obtusely difficult, especially creating payments like the Manual Purchases extension does.

We should really introduce an `EDD_Payment` class that takes care of this.

For example, to create a new payment, it might look like this:
```php
$payment = new EDD_Payment;
$payment->add_download( $id, $price_id );
$payment->add_fee( $label, $amount );
$payment->add_discount( $code );
$payment->set_total( $amount );
$payment->set_tax( $amount );
$payment->complete();
```
All meta data should be able to be set in this way.

Retrieving an existing payment would be as simple as:
```php
$payment = new EDD_Payment( $id );
$total = $payment->total;
$downloads = $payment->downloads;
$tax = $payment->tax;
```
Updating an existing payment:
```php
$payment = new EDD_Payment( $id );
$payment->set_total( $new_total );
$payment->add_note( 'Total manually changed because why not?' );
```

**Comment by ghost1227** [comment]
um... no...

**Comment by pippinsplugins** [comment]
Can you push it up? The commit links above all 404

**Comment by danieliser** [comment]
Any help needed on this?

**Comment by cklosowski** [comment]
Not yet. We're waiting on some work that's been done, then it's on my list to knock out for 2.3

**Comment by cklosowski** [comment]
@ghost1227 If you can, see what happened to your commits for this Would like to get things moving, If you can make it a PR or something that'd be great.

**Comment by pippinsplugins** [comment]
@ghost1227 If you can't find them, we'll rewrite it.

**Comment by ghost1227** [comment]
It's always been there... and I've linked it to you at least once in the last few weeks in hall. It's at https://github.com/easydigitaldownloads/Easy-Digital-Downloads/tree/feature/payments-class

**Comment by pippinsplugins** [comment]
Unfortunately we're running out of time for this so we're going to punt it to 2.4.

**Comment by ghost1227** [comment]
@pippinsplugins Anything that you see that needs to be fixed?

**Comment by pippinsplugins** [comment]
Since we have punted to 2.4, I haven't looked at it in depth yet. Will
after 2.3 is out.

**Comment by pippinsplugins** [comment]
`edd_insert_payment()` should function as a wrapper for this.

**Comment by pippinsplugins** [comment]
2.4 beta is Wednesday. We'll have to punt this to 2.5.

**Comment by cklosowski** [comment]
Blocker for #2714

**Comment by cklosowski** [comment]
This is in no way done, or ready for usage, just getting some changes pushed up for safe keeping. Right now i'm working through the updating of a payment via the Admin, and working out some kinks in the way removing and adding downloads from there integrates with the new class.

**Comment by pippinsplugins** [comment]
This is looking great!

**Comment by cklosowski** [comment]
Ok, I think this is ready for some eyes @pippinsplugins. I know for sure it's not 'done done', but I think I've taken it to the point where I can without some feedback. 

I've tested this out with Manual Purchases as well as Software Licensing, and I can't find anything 'in your face' breaking. Really testing extensions is just testing backwards compatibility, so we'll need to be sure we do a ton of that.

**Comment by chriscct7** [comment]
I'll test this on FES as well

**Comment by pippinsplugins** [comment]
@cklosowski This is looking really good. Here are some suggestions after a quick scan (haven't actually tested yet):

1. There are some inconsistencies with names for methods that update certain pieces of metadata:
- add_fee()
- add_discount()
- update_currency()
- set_gateway()
- add_note()
- update_status()

`update_currency()`, `set_gateway()`, and `update_status()`, however, should be standardized. I'd propose we use `update_*()` instead of `set_*()` since we also have `setup_*()` which could easily lead to confusion if we had both `set_*()` and `setup_*()`.

2. There are inconsistencies with some of the set/update methods. Some of them set the class properly only while others actually call the `update_meta()` method to save the change immediately. They should all only set the class property so there is no confusion on when `save()` should be called.

3. Many of our old payment functions have been updated to simply be a wrapper for the class methods. This is great but we need to make sure that the filters on the original payment functions are moved to the class methods as well so that all filters still run when interacting directly with the class and not the global functions.

For example, `edd_get_payment_amount()`. It has a filter of `edd_payment_amount` applied to the return value. This filter needs to be moved to `setup_total()` method in the class. That will make it so both of these behave the same way:

```php
$total = edd_get_payment_amount( $payment_id );

$payment = new EDD_Payment( $payment_id );
$total = $payment->total;
```

Some of the methods have had their filters properly moved. We just need to ensure all of them get moved.

4. In `edd_insert_payment()`, `do_action( 'edd_insert_payment', $payment->ID, $payment_data );` should fire only after `save()` has run successfully.

5. In `edd_update_payment_status()` both `status` and `post_status` are set. Can we make `post_status` get set automatically when running `save()` with a new status? Actually, this function is a bit weird. Calling `update_status()` should set the `status` and `post_status` properties so this should not be necessary:
```php
	if ( $updated ) {
		$payment->status         = $new_status;
		$payment->post_status    = $new_status;
		$payment->save();
	}
```

**Comment by pippinsplugins** [comment]
@cklosowski Let me know your thoughts on the above. I'm happy to start working through them as well but wanted to make sure we're both on the same page first.

**Comment by pippinsplugins** [comment]
Quick test shows that this all appears to be working really well.

**Comment by cklosowski** [comment]
A couple of thoughts on naming, yeah there are some inconsistencies, however there are a few things I'd like to keep in place:
Discounts, downloads, fees, notes, are all things we can have multiples of...so `add_` seems the best route for those.

The `set_` vs `update_` I'm good with moving towards `update_` I tried to keep all the filters in place in the `setup_` methods as those are called whenever our old `helper functions` are called due to the instantiation of the new class. So no matter what old helper function you call, the filters will still run during the `setup_` method. I'll make sure I got them all.

The handling of `status` and `post_status` was for backwards compatibility. I'll see about updating them better to be changed when updated, the problem is that the `__set()` magic doesn't get run when the class updates it's own properties, that's only for outside usage, so I'd have to be done in the `save()` method.

**Comment by pippinsplugins** [comment]
Personally I prefer `set_` over `update_`. I'd also love to see all set/update methods simply set the property and then have `save()` handle the actual update of the database. That way everything is consistent instead of some methods updating the DB and others not.

**Comment by cklosowski** [comment]
My concern with `set_` is the confusion with `setup_`, which are the ones run on instantiation to be sure all filters run.

I'm attempting to run it all items via the `save()`, the one you saw for the gateway was incorrectly in existence, prior to the `save()`.

**Comment by pippinsplugins** [comment]
That's a good point, but I don't think it's too big of a concern.

`update` inherently means "save changes", but we're not saving the changes until we call `save()`.

**Comment by cklosowski** [comment]
@pippinsplugins Ok, we've hit a headway. At this point we can move forward with discounts being part of the payment class, or not, but our current method of discounts doesn't apply cleanly to a new `EDD_Payment` class.

Do you mind jumping in and seeing how you feel about this, and working with it to see how you might add discount management to it?

**Comment by pippinsplugins** [comment]
@cklosowski Is the issue primarily with adding/removing discounts post creation?

**Comment by cklosowski** [comment]
yes, and how we handle `add_discount` if someone uses it in the pattern of:
```
$payment = new EDD_Payment;
$payment->create_payment( $data );
$payment->add_download(...);
$payment->add_discount(...);
$payment->save();
```

Do we even support that route? Do we throw an error, and only allow applying discounts to already created payments, etc.

**Comment by cklosowski** [comment]
the distinction here, is that Cart and Payment are getting VERY close to parity, but where do we draw that line?

**Comment by pippinsplugins** [comment]
I don't think we support that route at this time.

**Comment by pippinsplugins** [comment]
@cklosowski If you add a new download to a payment from `View Order Details`, this happens:
![screen shot 2015-11-27 at 2 02 51 pm](https://cloud.githubusercontent.com/assets/1034109/11447655/994abf58-950f-11e5-9457-b4e4324026e7.png)

**Comment by cklosowski** [comment]
Interesting it didn't a few commits ago I'll verify again.

**Comment by cklosowski** [comment]
Looks like debug info I left in.

**Comment by pippinsplugins** [comment]
Todo:

- [ ] Remove logging output when adding download to payment
- [ ] Update @since on main class file
- [ ] Add PHPDoc to methods that don't have it
- [ ] Remove `add_discount()` for now
- [ ]

Other thoughts:

1. Should we have a `update_currency()` method? Any reason not to just use `$payment->currency = 'USD'`?
2. Instead of having `setup_currency()` (and similar methods) take a `$payment_meta` parameter, could we have those pull from the object properties instead?

**Comment by cklosowski** [comment]
The setup methods are there to keep the old filters from running

**Comment by cklosowski** [comment]
Not "from running" but to keep them running.

**Comment by pippinsplugins** [comment]
@cklosowski Could you elaborate? I'm not seeing where `setup_currency()`, for example, triggers a filter.

**Comment by cklosowski** [comment]
Is it that function alone? The setups were there to maintain filters. Might have made that one thinking there was a filter in the old helper function.

**Comment by pippinsplugins** [comment]
@cklosowski We got confused. I was talking about `update_currency()` not `setup_currency()`. It's the `update_currency()` method that seems unnecessary.

**Comment by cklosowski** [comment]
Cool. Will fix that up.

**Comment by chriscct7** [comment]
Another really nice thing to have that would be easy to add is a "status_nicename" that converts the status to the actual display status, ex "publish" -> Complete, etc

**Comment by chriscct7** [comment]
It'd also be nice if there was a object key for Customer name that was set something like:
$customer_name  = isset( $order->user_info->first_name && isset( $order->user_info->last_name ) ) ? $order->user_info->first_name . ' ' . $order->user_info->last_name : 'Guest';

If we moved the payments table and the edit payments page to use the EDD_Payment objects, which would clean up a ton of code on the edit page in particular, then EDD core could use this key for the customer name column instead of relying on the post title or a customer object.

**Comment by pippinsplugins** [comment]
@cklosowski changes looking really good. A few comments:

1. I think we should consider removing `user_info` completely. It's a duplicate of the individual fields for the user now. Is it left just for the `edd_payment_meta_user_info` filter?

2. `$payment->gateway` is setup twice in `edd_insert_payment()`

3. All of the filters in the `setup_*()` methods need to pass `$this` to the filter so that the payment object can be referenced inside of callback functions. For filters that previously had a payment ID, keeps payment ID but then pass the object as the third parameter.

**Comment by cklosowski** [comment]
Sounds good, some feedback:
1) I think we should keep the user_info just set it to private. This way we don't have to rebuild that thing when we need to update it along with the actual DB entry. By setting it as private, it shouldn't be useful outside of internal usage.

2) Noted

3) Agreed.

**Comment by pippinsplugins** [comment]
1). Excellent idea.

**Comment by cklosowski** [comment]
@mindctrl This is ready for another go...

**Comment by mindctrl** [comment]
Just tested latest pull from here. I get a debug notice when making a purchase of a standard non-variable product:
`PHP Notice:  Undefined index: price_id in /wp-content/plugins/easy-digital-downloads/includes/payments/functions.php on line 128`

When using Manual Purchases (latest release, not the dev version updated for this), I get this also on a standard non-variable product:
`PHP Notice:  Undefined index: options in /wp-content/plugins/easy-digital-downloads/includes/payments/functions.php on line 128`

**Comment by mindctrl** [comment]
@pippinsplugins @cklosowski thoughts on making the `EDD_Payment` class `final` for a few releases until we've battle tested it, in case we need to change anything we're not anticipating right now?

**Comment by cklosowski** [comment]
@mindctrl @pippinsplugins I'm totally good with making it a `final` for a few releases until we've got this going.

**Comment by pippinsplugins** [comment]
Making it `final` sounds like a good plan to me as long as we remove it after 1-2 major versions.

**Comment by cklosowski** [comment]
I'll log the issue now for 2.6, and we can assess as we get there.

**Comment by pippinsplugins** [comment]
The date from Manual Purchases is not coming in correctly. It's always set to `January 1, 1970`.

**Comment by pippinsplugins** [comment]
Date works now.

**Comment by pippinsplugins** [comment]
There's something odd going on. I'm testing the creation of a simple payment like this:
```php
$payment = new EDD_Payment();
$payment->add_download( 97 );
$payment->save();
```
That should create a payment but it's not. It's bailing early somewhere.

**Comment by pippinsplugins** [comment]
Nevermind, I forgot that I had to set a customer name or email.

**Comment by pippinsplugins** [comment]
Works as expected:
```php
$payment = new EDD_Payment();
$payment->add_download( 97 );
$payment->email = 'pippin@pippinsplugins.com';
$payment->save();
```
Looks like IP and payment key are missing. We should set those if not set.
![screen shot 2015-12-01 at 11 51 34 am](https://cloud.githubusercontent.com/assets/1034109/11509143/fcb1a63c-9821-11e5-8a16-16472bd1941b.png)

**Comment by pippinsplugins** [comment]
Another item I just found.

If creating payments programmatically, customer records do not get attached. We currently have logic in `edd_insert_payment()` to attach a customer to a payment. This logic should be moved to `EDD_Payment()->insert_payment()`.

**Comment by pippinsplugins** [comment]
Another item.

Changing a payment from pending to complete programmatically did not generate the sale logs for the items.

Payment `218` was pending then I ran:
```php
$payment = new EDD_Payment( 218 );
$payment->status = 'complete';
$payment->save();
```
Payment was marked as complete:
![screen shot 2015-12-01 at 12 21 42 pm](https://cloud.githubusercontent.com/assets/1034109/11509839/1bbf995e-9826-11e5-9032-5be0af4e9082.png)
No sale log was created:
![screen shot 2015-12-01 at 12 21 57 pm](https://cloud.githubusercontent.com/assets/1034109/11509845/27e06c72-9826-11e5-9b6e-c62773ec764f.png)

**Comment by pippinsplugins** [comment]
Setting the payment to complete when first creating it did increase the sale count but no sale log entry was created:
```php
$payment = new EDD_Payment();
$payment->add_download( 97 );
$payment->email = 'pippin@pippinsplugins.com';
$payment->status = 'complete';
$payment->save();
```

**Comment by cklosowski** [comment]
@pippinsplugins ok i'll pull this down and work on the status update stuff

**Comment by cklosowski** [comment]
@pippinsplugins ok, so the issue we're hitting is that the log generation is in the part where the downloads are added to a payment...not on payment status change, let's have a quick chat about where we want this to exist.

**Comment by pippinsplugins** [comment]
How about just increasing it the same way we did before via the same actions. As long as the same actions are called, it should just work.

**Comment by pippinsplugins** [comment]
@cklosowski Looks like sale counts increase properly now!

**Comment by cklosowski** [comment]
Just moved the customer attachment to the `EDD_Payment` class, that way wether you use the class directly or still use `edd_insert_payment`, the customer creation is done and attached.

**Comment by pippinsplugins** [comment]
Looks like it works great!

**Comment by pippinsplugins** [comment]
This is merged into `release/2.5` but I'm leaving this open until we have further, real-world testing on this.

**Comment by chriscct7** [comment]
So just a thought, but making the class final broke an EDD experiment that
I'm working on which uses it. That's fine as I can hack my local EDD to
remove the final, but let's also keep in mind the same thing that
"protects" the class, is also preventing people who don't want to bother
hacking it from testing it.

On Thu, Dec 3, 2015 at 12:44 PM, Pippin Williamson <notifications@github.com

**Comment by cklosowski** [comment]
Final will come off this prior to 2.6. The final is so people don't build off it until we have it at a good place after some real world usage.

**Comment by pippinsplugins** [comment]
I'm seeing a notice on the payments list table:

NOTICE: wp-content/plugins/easy-digital-downloads/includes/admin/payments/class-payments-table.php:322 - Indirect modification of overloaded property EDD_Payment::$user_info has no effect
require_once('wp-admin/admin.php'), do_action('download_page_edd-payment-history'), call_user_func_array, edd_payment_history_page, WP_List_Table->display, WP_List_Table->display_rows_or_placeholder, WP_List_Table->display_rows, WP_List_Table->single_row, WP_List_Table->single_row_columns, call_user_func, EDD_Payment_History_Table->column_email

**Comment by pippinsplugins** [comment]
Including `parent` in the `$payment_data` array passed to `edd_insert_payment()` used to set the `post_parent` field. This appears to no longer work.

**Comment by pippinsplugins** [comment]
I'm finding a lot of notices when using `edd_insert_payment() without a full `$purchase_data` array. Going to commit fixes for each shortly.

**Comment by pippinsplugins** [comment]
Nevermind, all good. I was doing dumb things.

**Comment by pippinsplugins** [comment]
Looks like setting the status to `edd_subscription` results in the payment being created with a status of `pending`.

**Comment by pippinsplugins** [comment]
Status works now that `do_action( 'edd_insert_payment', $payment->ID, $payment_data );` runs later.

**Comment by pippinsplugins** [comment]
I'm seeing this notice again:

Notice: Indirect modification of overloaded property EDD_Payment::$user_info has no effect in /Users/pippin/Sites/edd/wp-content/plugins/easy-digital-downloads/includes/admin/payments/class-payments-table.php on line 322
![screen shot 2015-12-10 at 9 21 01 am](https://cloud.githubusercontent.com/assets/1034109/11719219/5b50112a-9f1f-11e5-951c-710e6bce2d9b.png)

**Comment by pippinsplugins** [comment]
Before we close this out, let's add PHPDoc to all properties.

**Comment by ghost1227** [comment]
@pippinsplugins I know it's not specific to this issue, but your comment made me think of it... Have you ever actually run apidoc on EDD? It reports 2010 errors in our docblocks. Hell, I've seen docblocks in EDD that are inaccurate but flag as valid too. We really need to clean them up at some point.

**Comment by pippinsplugins** [comment]
I have and .... to be super blunt, don't really care. It's technical debt that fixing has almost zero benefit.

**Comment by ghost1227** [comment]
If you really don't care, why is it important for a proper docblock to be added to this before release? I'm not suggesting that someone needs to do a full audit or anything, more that it would be worthwhile to look at the docblock for any function you're actively updating and ensure it's up to date. Doing a complete overhaul might be a tremendous waste of time, but improving individual docs as we change specific functions (or classes, etc) doesn't hurt anyone and results in a better, more valuable, and easier to understand codebase in the long run.

**Comment by pippinsplugins** [comment]
We are not going to devote time, money, and effort into going back and updating old ones. We *will* always do it properly moving forward.

**Comment by pippinsplugins** [comment]
@cklosowski Fees still aren't working properly.

@mindctrl and I found the problem when testing some issues with `EDD Wallet`, which uses fees.

After the payment is completed, the fee structure is different than it used to be.

To replicate:

1. checkout EDD `master`
2. Using Wallet, deposit funds into your wallet. It will work.
3. checkout EDD `release/2.5`
4. Using Wallet, deposit funds into your wallet. It will fail.

Here's how Wallet uses fees post-purchase:
```php
/**
 * Add the deposited amount to the wallet after payment
 *
 * @since       1.0.0
 * @param       int $payment_id The ID of the payment
 * @return      void
 */
function edd_wallet_add_funds( $payment_id ) {
	$fees = edd_get_payment_fees( $payment_id );

	if( $fees && count( $fees ) == 1 ) {
		if( $fees[0]['id'] == 'edd-wallet-deposit' ) {

			// Disable purchase receipts... we send our own emails
			remove_action( 'edd_complete_purchase', 'edd_trigger_purchase_receipt', 999 );

			// Send our custom emails
			edd_wallet_send_email( 'user', $payment_id );

			// Get the ID of the purchaser
			$user_id = edd_get_payment_user_id( $payment_id );

			// Deposit the funds
			edd_wallet()->wallet->deposit( $user_id, $fees[0]['amount'], 'deposit', $payment_id );

			// Tag the payment so we can find it later
			edd_update_payment_meta( $payment_id, '_edd_wallet_deposit', $user_id );
		}
	}
}
add_action( 'edd_complete_purchase', 'edd_wallet_add_funds' );

```

**Comment by pippinsplugins** [comment]
We've broken discount codes.

Payment records are created without discounted amounts. 

On checkout, this purchase showed as $8.00 after the 20% discount was applied: 
![screen shot 2015-12-22 at 3 52 45 pm](https://cloud.githubusercontent.com/assets/1034109/11966867/1a441dd0-a8c4-11e5-8f36-50a82380acb3.png)

**Comment by pippinsplugins** [comment]
There is a bug right now with the amounts / totals that get stored. I believe it's related to when there is a discount on individual items.

I had two $29.40 orders go through. The item amount and purchase total, however, got shown as $16.80.

The actual amount paid through merchant account was correct ($29.40).

@cklosowski

**Comment by pippinsplugins** [comment]
@cklosowski That change fixed it. Renewals are showing proper amounts now.

**Comment by phcaraujo** [comment]
@pippinsplugins 

Can you help me with this code?
I'm trying to create a purchase manually for product 3678

```
private function generate_payment($obj){
		$payment = new EDD_Payment();
		$payment->add_download( 3678 ); 
		$payment->email = $obj['email'];
		$payment->status = 'complete'; 
		$payment->save(); 
	}

    private function create_user($obj) {
        $obj['password'] = wp_generate_password(6, false);
        if (email_exists($obj["email"])) {
            //TODO: Melhorar esta abordagem, pois a intenção é mapear um reenvio do hotmart.
            // Talvez a solução seria quebrar em novos métodos e validar corretamente os possíveis cenários.
            $user = get_user_by( 'email', $obj["email"] );
            wp_set_password( $obj['password'], $user->ID );
            generate_payment($obj);            
        } else {
            $userdata = array(
                'user_login' => $obj['email'],
                'user_nicename' => $obj['first_name'],
                'first_name' => $obj['first_name'],
                'last_name' => $obj['last_name'],
                'user_email' => $obj['email'],
                'user_pass' => $obj['password'],
            );
            wp_insert_user($userdata);
			generate_payment($obj);
        }

        $this->send_email($obj);        
    }
```

This code is from another plugin that integrates my website with another platform.
